### PR TITLE
[SYCL] Optional kernel features: implement split based on reqd-sub-group-size

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
This patch implements device code split based on reqd-sub-group-size attribute, enables generation of "reqd_sub_group_size" property in "SYCL/device requirements" property set, and adds support of reqd_sub_group_size to sycl::is_compatible

Design:
https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md#changes-to-the-device-code-split-algorithm
E2E tests: https://github.com/intel/llvm-test-suite/pull/1569

### CODE REVIEW

File: llvm/test/tools/sycl-post-link/device-code-split/per-aspect-split-1.ll [hunk 1]
```diff
@@ -10,30 +10,30 @@
 ; RUN: sycl-post-link -split=auto -symbols -S %s -o %t.table
 ; RUN: FileCheck %s -input-file=%t_0.ll --check-prefixes CHECK-M0-IR \
 ; RUN: --implicit-check-not kernel0 --implicit-check-not kernel1
-; RUN: FileCheck %s -input-file=%t_1.ll --check-prefixes CHECK-M1-IR \
-; RUN: --implicit-check-not kernel0 --implicit-check-not kernel1
-; RUN: FileCheck %s -input-file=%t_2.ll --check-prefixes CHECK-M2-IR \
-; RUN: --implicit-check-not kernel0 --implicit-check-not kernel1
+; RUN: FileCheck %s -input-file=%t_1.ll --check-prefixes CHECK-M2-IR \
```
Comment (dm-vodopyanov) [inline llvm/test/tools/sycl-post-link/device-code-split/per-aspect-split-1.ll:13]: It is ok to change this test, per-reqd-wg-size-split-2.ll, sycl-esimd-large-grf.ll and sycl-large-grf.ll below: by adding new optional kernel feature reqd-sub-group-size to internal data structures, re-hashing happened which affected name generation logic. This does not affect customers as customers don't use these temp files directly, and didn't break anything in the pipeline.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 2]
```diff
@@ -2197,8 +2197,9 @@ bool doesDevSupportDeviceRequirements(const device &Dev,
 
   auto AspectsPropIt = getPropIt("aspects");
   auto ReqdWGSizePropIt = getPropIt("reqd_work_group_size");
+  auto ReqdSubGroupSizePropIt = getPropIt("reqd_sub_group_size");
 
-  if (!AspectsPropIt && !ReqdWGSizePropIt)
+  if (!AspectsPropIt && !ReqdWGSizePropIt && !ReqdSubGroupSizePropIt)
```
Comment (aelovikov-intel) [inline sycl/source/detail/program_manager/program_manager.cpp:2202]: nit: I'd remove these two lines alltogether.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 3]
```diff
@@ -2262,6 +2263,23 @@ bool doesDevSupportDeviceRequirements(const device &Dev,
           return false;
     }
   }
+
+  // Checking if device supports defined required sub-group size
```
Comment (aelovikov-intel) [inline sycl/source/detail/program_manager/program_manager.cpp:2267]:
```suggestion
  // Check if device supports required sub-group size.
```

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 4]
```diff
@@ -2262,6 +2263,23 @@ bool doesDevSupportDeviceRequirements(const device &Dev,
...
+        DeviceBinaryProperty(*(ReqdSubGroupSizePropIt.value())).asByteArray();
+    // Drop 8 bytes describing the size of the byte array.
+    ReqdSubGroupSize.dropBytes(8);
+    int ReqdSubGroupSizeVal = ReqdSubGroupSize.consume<int>();
+    auto SupportedSubGroupSizes = Dev.get_info<info::device::sub_group_sizes>();
+    if (!std::any_of(SupportedSubGroupSizes.cbegin(),
+                     SupportedSubGroupSizes.cend(),
+                     [&ReqdSubGroupSizeVal](int i) {
+                       return i == ReqdSubGroupSizeVal;
+                     })) {
```
Comment (aelovikov-intel) [inline sycl/source/detail/program_manager/program_manager.cpp:2279]:
```suggestion
    if (std::find(SupportedSubGroupSizes.cbegin(), SupportedSubGroupSizes.cend(),
                     ReqdSubGroupSizeVal) == SupportedSubGroupSizes.cend()) {
```

File: sycl/unittests/SYCL2020/IsCompatible.cpp [hunk 5]
```diff
@@ -69,13 +70,6 @@ template <> struct KernelInfo<TestKernelCPUInvalidReqdWGSize2D> {
   static constexpr int64_t getKernelSize() { return 1; }
 };
 
-} // namespace detail
-} // __SYCL_INLINE_VER_NAMESPACE(_V1)
-} // namespace sycl
-
-namespace sycl {
-__SYCL_INLINE_VER_NAMESPACE(_V1) {
-namespace detail {
```
Comment (aelovikov-intel) [inline sycl/unittests/SYCL2020/IsCompatible.cpp:78]: nit: Normally, I'd prefer those simplifications to be a separate PR.

File: sycl/unittests/helpers/PiImage.hpp [hunk 6]
```diff
@@ -505,13 +505,30 @@ inline PiProperty makeReqdWGSizeProp(const std::vector<int> &ReqdWGSize) {
...
+                          ReqdSubGroupSizePtr + sizeof(int),
+                          ValData.data() + BYTES_FOR_SIZE);
+  return {"reqd_sub_group_size", ValData, PI_PROPERTY_TYPE_BYTE_ARRAY};
+}
+
 inline void
 addDeviceRequirementsProps(PiPropertySet &Props,
                            const std::vector<sycl::aspect> &Aspects,
-                           const std::vector<int> &ReqdWGSize = {}) {
+                           const std::vector<int> &ReqdWGSize = {}, const int& ReqdSubGroupSize = 0) {
```
Comment (aelovikov-intel) [inline sycl/unittests/helpers/PiImage.hpp:526]:
```suggestion
                           const std::vector<int> &ReqdWGSize = {}, int ReqdSubGroupSize = 0) {
```

**Comment by dm-vodopyanov** [comment]
@v-klochkov [no-sycl-esimd-split-shared-func.ll](https://github.com/intel/llvm/blob/sycl/llvm/test/tools/sycl-post-link/sycl-esimd/no-sycl-esimd-split-shared-func.ll) test started to fail because this patch adds device code splitting for reqd_sub_group_size. Can you please advise, how to update this test? Or probably it doesn't make sense now?

```
FAIL: LLVM :: tools/sycl-post-link/sycl-esimd/no-sycl-esimd-split-shared-func.ll (44963 of 45993)
******************** TEST 'LLVM :: tools/sycl-post-link/sycl-esimd/no-sycl-esimd-split-shared-func.ll' FAILED ********************
Script:
--
: 'RUN: at line 8';   sycl-post-link -lower-esimd -symbols -split=auto -S /__w/llvm/llvm/src/llvm/test/tools/sycl-post-link/sycl-esimd/no-sycl-esimd-split-shared-func.ll -o /__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp.table
: 'RUN: at line 9';   /__w/llvm/llvm/build/bin/FileCheck /__w/llvm/llvm/src/llvm/test/tools/sycl-post-link/sycl-esimd/no-sycl-esimd-split-shared-func.ll -input-file=/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp.table --check-prefixes CHECK-TABLE
: 'RUN: at line 10';   /__w/llvm/llvm/build/bin/FileCheck /__w/llvm/llvm/src/llvm/test/tools/sycl-post-link/sycl-esimd/no-sycl-esimd-split-shared-func.ll -input-file=/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_0.ll
: 'RUN: at line 11';   /__w/llvm/llvm/build/bin/FileCheck /__w/llvm/llvm/src/llvm/test/tools/sycl-post-link/sycl-esimd/no-sycl-esimd-split-shared-func.ll -input-file=/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_0.sym --check-prefixes CHECK-SYM
--
Exit Code: 1

Command Output (stderr):
--
/__w/llvm/llvm/src/llvm/test/tools/sycl-post-link/sycl-esimd/no-sycl-esimd-split-shared-func.ll:16:21: error: CHECK-TABLE-EMPTY: is not on the line after the previous match
; CHECK-TABLE-EMPTY:
                    ^
/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp.table:5:1: note: 'next' match was here

^
/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp.table:2:330: note: previous match ended here
/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_0.ll|/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_0.prop|/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_0.sym
                                                                                                                                                                                                                                                                                                                                         ^
/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp.table:3:1: note: non-matching line after previous match is here
/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_1.ll|/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_1.prop|/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_1.sym
^

Input file: /__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp.table
Check file: /__w/llvm/llvm/src/llvm/test/tools/sycl-post-link/sycl-esimd/no-sycl-esimd-split-shared-func.ll

-dump-input=help explains the following input dump.

Input was:
<<<<<<
          1: [Code|Properties|Symbols] 
          2: /__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_0.ll|/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_0.prop|/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_0.sym 
          3: /__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_1.ll|/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_1.prop|/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_1.sym 
          4: /__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_esimd_2.ll|/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_esimd_2.prop|/__w/llvm/llvm/build/test/tools/sycl-post-link/sycl-esimd/Output/no-sycl-esimd-split-shared-func.ll.tmp_esimd_2.sym 
          5: 
empty:16     ! error: match on wrong line
>>>>>>

```

**Comment by dm-vodopyanov** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1569

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 2 above)
Comment (dm-vodopyanov) [reply sycl/source/detail/program_manager/program_manager.cpp]: Agree, this is redundant

**Comment by dm-vodopyanov** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1569

**Comment by v-klochkov** [comment]
> @v-klochkov [no-sycl-esimd-split-shared-func.ll](https://github.com/intel/llvm/blob/sycl/llvm/test/tools/sycl-post-link/sycl-esimd/no-sycl-esimd-split-shared-func.ll) test started to fail because this patch adds device code splitting for reqd_sub_group_size. Can you please advise, how to update this test? Or probably it doesn't make sense now?

I am not sure yet if the test should stay with your fix, 
but IMO the problems with this test showed some major concerns to this PR as the result doesn't look correct. with this fix/PR.

WITHOUT THE FIX:
a0) generated 1 output module
b0) SIMD_CALLEE got removed completely as it is alwaysinline
c0) SHARED_F is cloned: original SHARED_F() and new SHARED_F.esimd()
d0) SIMD_CALL_HELPER(func_ptr, 4xfloat) is transformed to SIMD_CALL_HELPER_1(4xfloat)


WITH THE FIX:
a1) Not OK: generated 3 modules - Half of ESIMD functions went to test_0.ll, another half in test_esimd2.ll
    In particular, this function got to test_0.ll instead of test_esimd2.ll:
	define dso_local spir_kernel void @ESIMD_kernel(i8 addrspace(1)* "VCArgumentIOKind"="0" %ptr) #5 !sycl_explicit_simd !1 !intel_reqd_sub_group_size !2 {

b1) Not OK: SIMD_CALLEE is not removed and left as a dead code in 
c1) Ok: SHARED_F is cloned: original SHARED_F() is in test_0.ll and new SHARED_F() is in test_esimd2.ll
d1) Not OK: SIMD_CALL_HELPER(func_ptr, 4xfloat) is transformed to SIMD_CALL_HELPER_1(4xfloat),
      but the declarations of SIMD_CALL_HELPER_1 in test_0.ll and test_1.ll lost the attribute "VCFunction".

**Comment by dm-vodopyanov** [comment]
>  IMO the problems with this test showed some major concerns to this PR as the result doesn't look correct. with this fix/PR.

Device code split based on reqd-sub-group-size is a part of optional kernel features which are part of SYCL 2020. This esimd test uses reqd-sub-group-size attribute, so this is expected that now more than one modules are generated because we do split on reqd-sub-group-size. So I guess test should be re-written or esimd feature should be revised to take into account this SYCL 2020 feature. E.g., regarding the test, removing ```-split=auto``` from the test makes test pass as no splitting is performed.

**Comment by v-klochkov** [comment]
> Device code split based on reqd-sub-group-size is a part of optional kernel features which are part of SYCL 2020

The attribute "!sycl_explicit_simd" must have higher priority than "!intel_reqd_sub_group_size 1".
I mean the functions with "!sycl_explicit_simd" must go to test_esimd_2.ll instead of test_0.ll.

It is Ok to separate funcs with "!intel_reqd_sub_group_size 1" to a separate module,
but those func should be SYCL (not ESIMD). All ESIMD funcs need to go to ESIMD module.

The current version of PR generates wrong IR.
In particular, test_0.ll has got ESIMD funcs (with !sycl_explicit_simd attr), but test_0.prop doesn't have "isEsimdImage=1|1".

The PR cloned SHARED_F 2 times instead of 1, which is unnecessary.

Also, why modified compiler moved this func to test_esimd_2.ll:
```
  define weak_odr dso_local x86_regcallcc <4 x float> @SIMD_CALL_HELPER_1(<4 x float> %simd_args) #3 !sycl_explicit_simd !0 !intel_reqd_sub_group_size !1
  !1 = !{i32 1}
```
and this one to test_0.ll? 
```
  define dso_local spir_kernel void @ESIMD_kernel(i8 addrspace(1)* "VCArgumentIOKind"="0" %ptr) #5 !sycl_explicit_simd !1 !intel_reqd_sub_group_size !2 {
  !2 = !{i32 1}
```

**Review by v-klochkov — CHANGES_REQUESTED** [review]
Mark as request-changes to finish discussion on ESIMD part before merging this.

**Comment by AlexeySachkov** [comment]
Superseded  by https://github.com/intel/llvm/pull/9928
